types/vote.go

```diff
@@ -410,33 +416,31 @@ func SignAndCheckVote(
 	v := vote.ToProto()
 	if err := privVal.SignVote(chainID, v); err != nil {
 		// Failing to sign a vote has always been a recoverable error, this function keeps it that way
-		return true, err // true = recoverable
+		return true, err
 	}
 	vote.Signature = v.Signature
+	vote.ExtensionSignature = nil
+	vote.Timestamp = v.Timestamp
 
-	isPrecommit := vote.Type == PrecommitType
-	if !isPrecommit && extensionsEnabled {
-		// Non-recoverable because the caller passed parameters that don't make sense
-		return false, fmt.Errorf("only Precommit votes may have extensions enabled; vote type: %d", vote.Type)
-	}
-
-	isNil := vote.BlockID.IsNil()
-	extSignature := (len(v.ExtensionSignature) > 0)
-	if extSignature == (!isPrecommit || isNil) {
-		// Non-recoverable because the vote is malformed
-		return false, fmt.Errorf(
-			"extensions must be present IFF vote is a non-nil Precommit; present %t, vote type %d, is nil %t",
-			extSignature,
-			vote.Type,
-			isNil,
+	if extensionsEnabled {
+		var (
+			extSignature = (len(v.ExtensionSignature) > 0)
+			isPrecommit  = (vote.Type == PrecommitType)
+			isNil        = vote.BlockID.IsNil()
 		)
-	}
 
-	vote.ExtensionSignature = nil
-	if extensionsEnabled {
+		if !isPrecommit {
+			// Non-recoverable because the vote is malformed.
+			panic(fmt.Sprintf("extensions are enabled, but vote is not Precommit (vote type: %T)", vote.Type))
+		}
+
+		if extSignature && isNil {
```

Comment:
From the logic here, if the Precommit is for nil but the signature is empty, we are good.

Not sure how it should work in this case.

Reply:
From the contract, we should panic here.